### ISSUE TITLE
[FEATURE]

### ISSUE THREAD

**Issue by preethikalyani88** [opensearch-project/index-management#399]
**Is your feature request related to a problem?**
Open search index transforms as it is today only provides the ability to update only transformed buckets if it has data modifications. It does not handle the case where deletion of documents/Modifications  to documents in the source index can result in deletion of a transformed bucket in target index as it is. This is a really needed feature when source index documents can be deleted or updated such that it obviates the bucket in transformed index as it is. 
For example:
With a sample document like this :

```
{
  "CustId": "000",
  "status": "ACTIVE"
}
```
And transform job like this 

```
transform:{
  ......
  .....
  groups: [
        {
            "terms":
            {
                   "source_field": "status.keyword"
            }
        }
  ], 
  aggregations: {
    "count": {
        "value_count": {
              "field": "CustId"
         }
     }
  }
}
```
The initial transform run does the count correctly as follows in the targeted index:
```
{
  "count": "1.0",
  "status": "ACTIVE"
}
```
And if the source document's STATUS is modified to CLOSED from ACTIVE, the transformed index now has two documents as follows:
```
{
  "count": "1.0",
  "status": "ACTIVE"
},
{
  "count": "1.0",
  "status": "CLOSED"
}
```

It does not remove the active bucket after adding the closed bucket. 

**What solution would you like?**
We want the targeted index to have documents as follows: 

```
{
  "count": "0.0",
  "status": "ACTIVE"
}
```
Or just remove the bucket fully as well. 
Clean up being async than the default transform result is also an agreeable solution for us as our system is eventually consistent anyways. 

**What alternatives have you considered?**

As an alternative to to ensure that we can achieve similar results we have modified source index schema as follows:
```
{
  "CustId": "000",
  "ACTIVE": 1.0
}
```
This is fine as long as cardinality of counted fields are smaller but in other cases it makes it impossible.

**Comment by downsrob** [comment]
First, I'll explain how continuous transforms works, and then I will try to clarify why it is difficult to delete buckets.
Continuous mode transforms track the sequence number checkpoints at each run, then perform the transform's normal composite aggregation on only the new documents with no metrics included in the aggregation. We use this to get just the set of modified buckets from the last run. We then perform another composite aggregation on this set of buckets using all of the documents in the source index. The result is that only the buckets modified since the last run are indexed.

Deleting stale buckets is difficult because we don't have any method to query for the documents which were deleted; there is no history in OpenSearch. I am always open to a clever solution, but to find the deleted buckets I believe that we would need to perform a full transform from the source index to a new index, and then find the bucket difference between the indices, and delete these from the transform target index. This is very inefficient. Some core optimizations might be to listen to all document deletions for the transform source index and make a note of which buckets need to be rerun due to deletions, but we have found that doing such a heavy operation pipelined into document deletions has a really heavy load as well.

One option is a separate feature to simply delete old documents in a target index at regular intervals. By introducing a new scheduled job to take note of the current sequence number checkpoint in an index, we could then delete the documents with a lower sequence number after x time has passed since the sequence number was recorded. Delete by query is an expensive operation, however, and for continuous transforms this would not distinguish between a bucket which has not been updated but is still valid, and a bucket which was deleted.

**Comment by preethikalyani88** [comment]
Yeah, this is the key reason: "for continuous transforms this would not distinguish between a bucket which has not been updated but is still valid, and a bucket which was deleted"  why I think this option is harder. So on the same note, it seems the crux is really the checkpoints right ? Checkpoints literally seems like a snapshot , do you forsee a case where you can find the difference between two checkpoints ?

**Comment by rishabhmaurya** [comment]
We should explore the use shard history retention lease and treat transform index shards as one of its consumers.
